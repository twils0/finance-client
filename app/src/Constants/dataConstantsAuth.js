export const actionTypes = {
  SET_AUTHENTICATED: 'DATA_AUTH_SET_AUTHENTICATED',
  SET_REDIRECT_URL: 'DATA_AUTH_SET_REDIRECT_URL',
  SET_AUTH_STATUS: 'DATA_AUTH_SET_AUTH_STATUS',
  SET_CODE_TYPE: 'DATA_AUTH_SET_CODE_TYPE',
  RESET_AUTH_STATE: 'DATA_AUTH_RESET_AUTH_STATE',
};

export const statusNames = {
  LOGIN: 'authStatusLogin',
  SIGN_OUT_DEVICES: 'authStatusSignOutDevices',
  LOGIN_MFA: 'authStatusLoginMFA',
  SIGN_UP: 'authStatusSignUp',
  VERIFY_PHONE: 'authStatusVerifyPhone',
  VERIFY_PHONE_CODE: 'authStatusVerifyPhoneCode',
  VERIFY_EMAIL: 'authStatusVerifyEmail',
  VERIFY_EMAIL_LINK: 'authStatusVerifyEmailLink',
  FORGOT_PASSWORD: 'authStatusForgotPassword',
  RESET_PASSWORD: 'authStatusResetPassword',
  CHANGE_PASSWORD: 'authStatusChangePassword',
  DELETE_ACCOUNT: 'authStatusDeleteAccount',
  LOGOUT: 'authStatusLogout',
};

export const codeTypeNames = {
  VERIFY_PHONE: 'verifyPhone',
  VERIFY_EMAIL: 'verifyEmail',
  VERIFY_EMAIL_ADDITIONAL: 'verifyEmailAdditional',
};
