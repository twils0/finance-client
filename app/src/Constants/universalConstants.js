import stripeConfig from '../../../stripe.config.json';
import dbConfig from '../../../db.config.json';
import iexConfig from '../../../iex.config.json';

export const baseURLs = {
  STRIPE: stripeConfig.url,
  DB: dbConfig.url,
  IEX: iexConfig.urls.api,
  IEX_DEV: iexConfig.urls.dev,
  IEX_TERMS: iexConfig.urls.terms,
};

export const URLs = {
  CUSTOMERS: '/customers',
  USERS: '/users',
  EMAILS: '/emails',
  SECURITIES: '/securities',
  SECURITIES_SEARCH: '/securities/search',
  IEX_COMPANY: tickerCusip => `/stock/${tickerCusip}/company`,
  IEX_QUOTE: tickerCusip => `/stock/${tickerCusip}/quote`,
};

export const axiosConfig = {
  STRIPE: {
    baseURL: baseURLs.STRIPE,
    timeout: 30000,
  },
  DB: {
    baseURL: baseURLs.DB,
    timeout: 30000,
  },
  IEX: {
    baseURL: baseURLs.IEX,
    timeout: 30000,
  },
};

export const pathNames = {
  DEFAULT: '*',
  LOGIN: '/login',
  SIGN_UP: '/signup',
  CODE: '/code',
  CODE_VERIFY_EMAIL: '/code/:verificationId?',
  TERMS: '/terms',
  INTERNAL: '/internal',
  WATCHLIST: '/internal/watchlist',
  WATCHLIST_SECURITY_ID: '/internal/watchlist/:securityId?',
  ACCOUNT: '/internal/account',
};

export const requestStatusTypes = {
  IDLE: 'IDLE',
  LOADING: 'LOADING',
  SUCCESS: 'SUCCESS',
  ERROR: 'ERROR',
};
