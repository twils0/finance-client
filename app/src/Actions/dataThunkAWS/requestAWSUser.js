import { Auth } from 'aws-amplify';
import raven from 'raven-js';

import handleErrorCatch from '../../handleErrorCatch';
import { requestStatusTypes } from '../../Constants/universalConstants';
import { setAWSStatus, setAWSUser } from '../dataActionsAWS';
import requestLogout from '../dataThunkAuth/requestLogout';

import { demo } from '../../../../mode.config.json';

const requestAWSUser = () => async (dispatch, getState) => {
  const state = getState();

  if (state.data.aws.status !== requestStatusTypes.LOADING) {
    dispatch(setAWSStatus({ status: requestStatusTypes.LOADING }));

    let user = null;

    try {
      if (!demo) {
        user = await Auth.currentUserPoolUser();
      }
    } catch (errorCatch) {
      const error = handleErrorCatch(errorCatch);

      dispatch(setAWSStatus({ status: requestStatusTypes.ERROR }));

      if (
        (typeof error === 'string'
          && (error === 'not authenticated' || error === 'No current user in userPool'))
        || (typeof error === 'object' && error && error.code === 'NotAuthorizedException')
      ) {
        dispatch(requestLogout());

        return null;
      }

      if (!demo) {
        raven.captureException(error, {
          logger: 'requestAWSUser',
        });
      }

      return Promise.reject(error);
    }

    if (!demo) {
      dispatch(setAWSUser({ user }));
    }
    dispatch(setAWSStatus({ status: requestStatusTypes.SUCCESS }));

    return { user };
  }

  return null;
};

export default requestAWSUser;
