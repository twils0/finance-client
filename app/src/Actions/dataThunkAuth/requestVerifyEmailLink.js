import axios from 'axios';
import raven from 'raven-js';

import { requestStatusTypes, URLs, axiosConfig } from '../../Constants/universalConstants';
import { statusNames } from '../../Constants/dataConstantsAuth';
import { setAuthStatus } from '../dataActionsAuth';

import handleErrorCatch from '../../handleErrorCatch';

import { demo } from '../../../../mode.config.json';

const requestVerifyEmailLink = (payload) => {
  if (!Object.prototype.hasOwnProperty.call(payload, 'verificationId')) {
    throw Error(`Please enter a value for the 'verificationId' key - ${payload}`);
  }

  return async (dispatch, getState) => {
    const state = getState();

    if (
      state.data.auth.status[statusNames.VERIFY_EMAIL_LINK].status !== requestStatusTypes.LOADING
    ) {
      dispatch(
        setAuthStatus({
          id: statusNames.VERIFY_EMAIL_LINK,
          status: requestStatusTypes.LOADING,
        }),
      );

      const { verificationId } = payload;

      try {
        if (!demo) {
          await axios.put(URLs.EMAILS, { verificationId }, axiosConfig.DB);
        }
      } catch (errorCatch) {
        const error = handleErrorCatch(errorCatch);

        dispatch(
          setAuthStatus({
            id: statusNames.VERIFY_EMAIL_LINK,
            status: requestStatusTypes.ERROR,
          }),
        );

        if (!demo) {
          raven.captureException(error, {
            logger: 'requestVerifyEmailLink',
          });
        }

        return Promise.reject(error);
      }

      dispatch(
        setAuthStatus({
          id: statusNames.VERIFY_EMAIL_LINK,
          status: requestStatusTypes.SUCCESS,
        }),
      );
    }

    return null;
  };
};

export default requestVerifyEmailLink;
