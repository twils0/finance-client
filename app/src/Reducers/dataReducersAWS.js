import { combineReducers } from 'redux';

import { requestStatusTypes } from '../Constants/universalConstants';
import { actionTypes } from '../Constants/dataConstantsAWS';

export const initialState = {
  status: requestStatusTypes.IDLE,
  user: null,
};

export const status = (state = initialState.status, action) => {
  switch (action.type) {
    case actionTypes.SET_AWS_STATUS:
      return action.payload.status;
    case actionTypes.RESET_AWS_STATE:
      return initialState.status;
    default:
      return state;
  }
};

export const user = (state = initialState.user, action) => {
  switch (action.type) {
    case actionTypes.SET_AWS_USER:
      return action.payload.user;
    case actionTypes.RESET_AWS_STATE:
      return initialState.user;
    default:
      return state;
  }
};

const dataReducersAWS = combineReducers({
  status,
  user,
});

export default dataReducersAWS;
