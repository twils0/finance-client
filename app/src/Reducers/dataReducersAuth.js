import { combineReducers } from 'redux';

import { requestStatusTypes, pathNames } from '../Constants/universalConstants';
import { actionTypes, statusNames, codeTypeNames } from '../Constants/dataConstantsAuth';

export const initialState = {
  authenticated: false,
  redirectURL: pathNames.WATCHLIST,
  status: {
    [statusNames.LOGIN]: {
      id: statusNames.LOGIN,
      status: requestStatusTypes.IDLE,
    },
    [statusNames.SIGN_OUT_DEVICES]: {
      id: statusNames.SIGN_OUT_DEVICES,
      status: requestStatusTypes.IDLE,
    },
    [statusNames.LOGIN_MFA]: {
      id: statusNames.LOGIN_MFA,
      status: requestStatusTypes.IDLE,
    },
    [statusNames.SIGN_UP]: {
      id: statusNames.SIGN_UP,
      status: requestStatusTypes.IDLE,
    },
    [statusNames.VERIFY_PHONE]: {
      id: statusNames.VERIFY_PHONE,
      status: requestStatusTypes.IDLE,
    },
    [statusNames.VERIFY_PHONE_CODE]: {
      id: statusNames.VERIFY_PHONE_CODE,
      status: requestStatusTypes.IDLE,
    },
    [statusNames.VERIFY_EMAIL]: {
      id: statusNames.VERIFY_EMAIL,
      status: requestStatusTypes.IDLE,
    },
    [statusNames.VERIFY_EMAIL_LINK]: {
      id: statusNames.VERIFY_EMAIL_LINK,
      status: requestStatusTypes.IDLE,
    },
    [statusNames.FORGOT_PASSWORD]: {
      id: statusNames.FORGOT_PASSWORD,
      status: requestStatusTypes.IDLE,
    },
    [statusNames.RESET_PASSWORD]: {
      id: statusNames.RESET_PASSWORD,
      status: requestStatusTypes.IDLE,
    },
    [statusNames.CHANGE_PASSWORD]: {
      id: statusNames.CHANGE_PASSWORD,
      status: requestStatusTypes.IDLE,
    },
    [statusNames.LOGOUT]: {
      id: statusNames.LOGOUT,
      status: requestStatusTypes.SUCCESS,
    },
    [statusNames.DELETE_ACCOUNT]: {
      id: statusNames.DELETE_ACCOUNT,
      status: requestStatusTypes.SUCCESS,
    },
  },
  codeTypes: {
    [codeTypeNames.VERIFY_PHONE]: {
      id: codeTypeNames.VERIFY_PHONE,
      needed: true,
    },
    [codeTypeNames.VERIFY_EMAIL]: {
      id: codeTypeNames.VERIFY_EMAIL,
      needed: true,
    },
    [codeTypeNames.VERIFY_EMAIL_ADDITIONAL]: {
      id: codeTypeNames.VERIFY_EMAIL_ADDITIONAL,
      needed: true,
    },
  },
};

export const authenticated = (state = initialState.authenticated, action) => {
  switch (action.type) {
    case actionTypes.SET_AUTHENTICATED:
      return action.payload.authenticated;
    case actionTypes.RESET_AUTH_STATE:
      return initialState.authenticated;
    default:
      return state;
  }
};

export const redirectURL = (state = initialState.redirectURL, action) => {
  switch (action.type) {
    case actionTypes.SET_REDIRECT_URL:
      return action.payload.redirectURL;
    default:
      return state;
  }
};

export const status = (state = initialState.status, action) => {
  switch (action.type) {
    case actionTypes.SET_AUTH_STATUS:
      return {
        ...state,
        [action.payload.id]: {
          ...state[action.payload.id],
          ...action.payload,
        },
      };
    case actionTypes.RESET_AUTH_STATE:
      return initialState.status;
    default:
      return state;
  }
};

export const codeTypes = (state = initialState.codeTypes, action) => {
  switch (action.type) {
    case actionTypes.SET_CODE_TYPE:
      return {
        ...state,
        [action.payload.id]: {
          ...state[action.payload.id],
          ...action.payload,
        },
      };
    case actionTypes.RESET_AUTH_STATE:
      return initialState.codeTypes;
    default:
      return state;
  }
};

const dataReducersAuth = combineReducers({
  authenticated,
  redirectURL,
  status,
  codeTypes,
});

export default dataReducersAuth;
