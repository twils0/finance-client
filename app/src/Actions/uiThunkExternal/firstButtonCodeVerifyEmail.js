import { requestStatusTypes, pathNames } from '../../Constants/universalConstants';
import { statusNames } from '../../Constants/dataConstantsAuth';
import { formNames, inputNames } from '../../Constants/uiConstantsExternal';
import { setAuthenticated } from '../dataActionsAuth';
import requestVerifyEmailLink from '../dataThunkAuth/requestVerifyEmailLink';
import requestAWSUser from '../dataThunkAWS/requestAWSUser';
import { setCurrentForm, setInputValueError } from '../uiActionsExternal';

const firstButtonCodeEmail = (payload) => {
  if (!Object.prototype.hasOwnProperty.call(payload, 'match')) {
    throw new Error(`Please enter a value for the 'match' key - ${JSON.stringify(payload)}`);
  }

  return async (dispatch, getState) => {
    const state = getState();
    let login = false;
    const { verificationId } = payload.match.params;
    const { status } = state.data.auth;

    try {
      if (verificationId) {
        let { user } = state.data.aws;

        if (!user) {
          ({ user } = await dispatch(requestAWSUser()));
        }

        await dispatch(requestVerifyEmailLink({ verificationId }));

        if (user) {
          login = true;
        } else {
          payload.history.replace(pathNames.LOGIN);
        }
      } else if (status[statusNames.LOGIN_MFA].status === requestStatusTypes.SUCCESS) {
        dispatch(setCurrentForm({ current: formNames.DEVICE }));
      } else {
        login = true;
      }

      if (login) {
        const { forms } = state.ui.external;
        const loginEmail = forms[formNames.LOGIN].inputs[inputNames[formNames.LOGIN].EMAIL];
        const loginPassword = forms[formNames.LOGIN].inputs[inputNames[formNames.LOGIN].PASSWORD];
        const currentTime = new Date();

        currentTime.setMinutes(currentTime.getMinutes() + 30);
        window.sessionStorage.setSecurity('sessionTime', currentTime);

        dispatch(setAuthenticated({ authenticated: true }));

        if (loginEmail.value || loginEmail.errorMessage) {
          loginEmail.value = '';
          loginEmail.errorMessage = '';

          dispatch(setInputValueError(loginEmail));
        }

        if (loginPassword.value || loginPassword.errorMessage) {
          loginPassword.value = '';
          loginPassword.errorMessage = '';

          dispatch(setInputValueError(loginPassword));
        }
      }
    } catch (errorCatch) {
      payload.history.replace(pathNames.LOGIN);
    }

    return null;
  };
};

export default firstButtonCodeEmail;
