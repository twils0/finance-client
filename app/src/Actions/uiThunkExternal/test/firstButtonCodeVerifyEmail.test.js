/**
 * @jest-environment node
 */

import configureMockStore from 'redux-mock-store';
import thunk from 'redux-thunk';
import MockDate from 'mockdate';

import { initialState as initialStateAuth } from '../../../Reducers/dataReducersAuth';
import { initialState as initialStateAWS } from '../../../Reducers/dataReducersAWS';
import { initialState as initialStateUIExternal } from '../../../Reducers/uiReducersExternal';
import { requestStatusTypes, pathNames } from '../../../Constants/universalConstants';
import { actionTypes as actionTypesAuth, statusNames } from '../../../Constants/dataConstantsAuth';
import {
  actionTypes as actionTypesUIExternal,
  formNames,
  inputNames,
} from '../../../Constants/uiConstantsExternal';
import requestVerifyEmailLink from '../../dataThunkAuth/requestVerifyEmailLink';
import requestAWSUser from '../../dataThunkAWS/requestAWSUser';
import firstButtonCodeVerifyEmail from '../firstButtonCodeVerifyEmail';

const history = {
  replace: jest.fn(),
};

const setSecurity = jest.fn();

global.window = {
  sessionStorage: {
    setSecurity,
  },
};
jest.mock('../../dataThunkAuth/requestVerifyEmailLink', () => jest.fn());
jest.mock('../../dataThunkAWS/requestAWSUser', () => jest.fn());

const middleware = [thunk];
const mockStore = configureMockStore(middleware);

const verificationId = '1234-5678-1234-5678';
const user = { test: 'testUser' };

const match = {
  params: {
    verificationId,
  },
};

describe('uiThunkExternal', () => {
  describe('firstButtonCodeVerifyEmail', () => {
    afterEach(() => {
      requestVerifyEmailLink.mockReset();
      requestAWSUser.mockReset();
      history.replace.mockReset();
      setSecurity.mockReset();
      MockDate.reset();
    });

    it("fails and throws an error when missing 'history' key in payload", async () => {
      const wrongPayload = {};

      const store = mockStore({
        data: { auth: initialStateAuth, aws: initialStateAWS },
        ui: { external: initialStateUIExternal },
      });

      const expectedActions = [];

      try {
        await store.dispatch(firstButtonCodeVerifyEmail(wrongPayload));
      } catch (errorCatch) {
        expect(errorCatch).toMatchSnapshot();
      }

      const actions = store.getActions();

      expect(actions).toEqual(expectedActions);
    });

    it("fails and throws an error when missing 'match' key in payload", async () => {
      const wrongPayload = {
        history,
      };

      const store = mockStore({
        data: { auth: initialStateAuth, aws: initialStateAWS },
        ui: { external: initialStateUIExternal },
      });

      const expectedActions = [];

      try {
        await store.dispatch(firstButtonCodeVerifyEmail(wrongPayload));
      } catch (errorCatch) {
        expect(errorCatch).toMatchSnapshot();
      }

      const actions = store.getActions();

      expect(actions).toEqual(expectedActions);
    });

    it('creates the correct actions with the correct payload, with user and verification id', async () => {
      const stateBeforeAWS = JSON.parse(JSON.stringify(initialStateAWS));

      stateBeforeAWS.user = user;

      const store = mockStore({
        data: { auth: initialStateAuth, aws: stateBeforeAWS },
        ui: { external: initialStateUIExternal },
      });

      const expectedActions = [
        {
          type: actionTypesAuth.SET_AUTHENTICATED,
          payload: {
            authenticated: true,
          },
        },
      ];

      const startDate = new Date('1/2/2000');
      const incrementDate = new Date('1/2/2000');
      const expectDate = new Date(incrementDate.setMinutes(incrementDate.getMinutes() + 30));

      MockDate.set(startDate);

      requestVerifyEmailLink.mockReturnValue(() => Promise.resolve(null));

      const result = await store.dispatch(firstButtonCodeVerifyEmail({ history, match }));

      const actions = store.getActions();

      expect(actions).toEqual(expectedActions);
      expect(result).toEqual(null);
      expect(history.replace).not.toBeCalled();
      expect(requestVerifyEmailLink).toBeCalledWith({ verificationId });
      expect(setSecurity).toBeCalledWith('sessionTime', expectDate);
    });

    it('creates the correct actions with the correct payload, with user from requestAWSUser and verification id', async () => {
      const store = mockStore({
        data: { auth: initialStateAuth, aws: initialStateAWS },
        ui: { external: initialStateUIExternal },
      });

      const expectedActions = [
        {
          type: actionTypesAuth.SET_AUTHENTICATED,
          payload: {
            authenticated: true,
          },
        },
      ];

      const startDate = new Date('1/2/2000');
      const incrementDate = new Date('1/2/2000');
      const expectDate = new Date(incrementDate.setMinutes(incrementDate.getMinutes() + 30));

      MockDate.set(startDate);

      requestAWSUser.mockReturnValue(() => Promise.resolve({ user }));
      requestVerifyEmailLink.mockReturnValue(() => Promise.resolve(null));

      const result = await store.dispatch(firstButtonCodeVerifyEmail({ history, match }));

      const actions = store.getActions();

      expect(actions).toEqual(expectedActions);
      expect(result).toEqual(null);
      expect(requestAWSUser).toBeCalled();
      expect(history.replace).not.toBeCalled();
      expect(requestVerifyEmailLink).toBeCalledWith({ verificationId });
      expect(setSecurity).toBeCalledWith('sessionTime', expectDate);
    });

    it('creates the correct actions with the correct payload, without user and with verification id', async () => {
      const store = mockStore({
        data: { auth: initialStateAuth, aws: initialStateAWS },
        ui: { external: initialStateUIExternal },
      });

      const expectedActions = [];

      requestAWSUser.mockReturnValue(() => Promise.resolve({}));
      requestVerifyEmailLink.mockReturnValue(() => Promise.resolve(null));

      const result = await store.dispatch(firstButtonCodeVerifyEmail({ history, match }));

      const actions = store.getActions();

      expect(actions).toEqual(expectedActions);
      expect(result).toEqual(null);
      expect(history.replace).toBeCalledWith(pathNames.LOGIN);
      expect(requestVerifyEmailLink).toBeCalledWith({ verificationId });
      expect(setSecurity).not.toBeCalled();
    });

    it('requestAWSUser throws an unexpected error, redirects to login', async () => {
      const stateBeforeAuth = JSON.parse(JSON.stringify(initialStateAuth));

      stateBeforeAuth.status[statusNames.LOGIN_MFA].status = requestStatusTypes.SUCCESS;

      const store = mockStore({
        data: { auth: stateBeforeAuth, aws: initialStateAWS },
        ui: { external: initialStateUIExternal },
      });

      const expectedActions = [];

      requestAWSUser.mockReturnValue(() => Promise.reject(null));

      const result = await store.dispatch(firstButtonCodeVerifyEmail({ history, match }));

      const actions = store.getActions();

      expect(actions).toEqual(expectedActions);
      expect(result).toEqual(null);
      expect(requestVerifyEmailLink).not.toBeCalled();
      expect(history.replace).toBeCalledWith(pathNames.LOGIN);
      expect(setSecurity).not.toBeCalled();
    });

    it('requestVerifyEmailLink throws an unexpected error, redirects to login', async () => {
      const stateBeforeAWS = JSON.parse(JSON.stringify(initialStateAWS));

      stateBeforeAWS.user = user;

      const store = mockStore({
        data: { auth: initialStateAuth, aws: stateBeforeAWS },
        ui: { external: initialStateUIExternal },
      });

      const expectedActions = [];

      requestAWSUser.mockReturnValue(() => Promise.resolve({ user }));
      requestVerifyEmailLink.mockReturnValue(() => Promise.reject(null));

      const result = await store.dispatch(firstButtonCodeVerifyEmail({ history, match }));

      const actions = store.getActions();

      expect(actions).toEqual(expectedActions);
      expect(result).toEqual(null);
      expect(requestVerifyEmailLink).toBeCalledWith({ verificationId });
      expect(history.replace).toBeCalledWith(pathNames.LOGIN);
      expect(setSecurity).not.toBeCalled();
    });

    it('creates the correct actions with the correct payload, with no verification id and login mfa status equals success', async () => {
      const stateBeforeAuth = JSON.parse(JSON.stringify(initialStateAuth));

      stateBeforeAuth.status[statusNames.LOGIN_MFA].status = requestStatusTypes.SUCCESS;

      const store = mockStore({
        data: { auth: stateBeforeAuth, aws: initialStateAWS },
        ui: { external: initialStateUIExternal },
      });

      const expectedActions = [
        {
          type: actionTypesUIExternal.SET_CURRENT_FORM,
          payload: {
            current: formNames.DEVICE,
          },
        },
      ];

      requestVerifyEmailLink.mockReturnValue(() => Promise.resolve(null));

      const result = await store.dispatch(
        firstButtonCodeVerifyEmail({ history, match: { params: {} } }),
      );

      const actions = store.getActions();

      expect(actions).toEqual(expectedActions);
      expect(result).toEqual(null);
      expect(history.replace).not.toBeCalled();
      expect(requestVerifyEmailLink).not.toBeCalled();
      expect(setSecurity).not.toBeCalled();
    });

    it('creates the correct actions with the correct payload, with no verification id and login mfa status not equal to success', async () => {
      const stateBeforeUIExternal = JSON.parse(JSON.stringify(initialStateUIExternal));

      stateBeforeUIExternal.forms[formNames.LOGIN].inputs[inputNames[formNames.LOGIN].EMAIL].value = 'testValue';
      stateBeforeUIExternal.forms[formNames.LOGIN].inputs[
        inputNames[formNames.LOGIN].PASSWORD
      ].value = 'testValue';

      const loginEmail = stateBeforeUIExternal.forms[formNames.LOGIN].inputs[inputNames[formNames.LOGIN].EMAIL];
      const loginPassword = stateBeforeUIExternal.forms[formNames.LOGIN].inputs[inputNames[formNames.LOGIN].PASSWORD];

      const payloadLoginEmail = JSON.parse(JSON.stringify(loginEmail));
      const payloadLoginPassword = JSON.parse(JSON.stringify(loginPassword));

      payloadLoginEmail.value = '';
      payloadLoginPassword.value = '';

      const store = mockStore({
        data: { auth: initialStateAuth, aws: initialStateAWS },
        ui: { external: stateBeforeUIExternal },
      });

      const expectedActions = [
        {
          type: actionTypesAuth.SET_AUTHENTICATED,
          payload: {
            authenticated: true,
          },
        },
        {
          type: actionTypesUIExternal.SET_INPUT_VALUE_ERROR,
          payload: payloadLoginEmail,
        },
        {
          type: actionTypesUIExternal.SET_INPUT_VALUE_ERROR,
          payload: payloadLoginPassword,
        },
      ];

      requestVerifyEmailLink.mockReturnValue(() => Promise.resolve(null));

      const startDate = new Date('1/2/2000');
      const incrementDate = new Date('1/2/2000');
      const expectDate = new Date(incrementDate.setMinutes(incrementDate.getMinutes() + 30));

      MockDate.set(startDate);

      const result = await store.dispatch(
        firstButtonCodeVerifyEmail({ history, match: { params: {} } }),
      );

      const actions = store.getActions();

      expect(actions).toEqual(expectedActions);
      expect(result).toEqual(null);
      expect(history.replace).not.toBeCalled();
      expect(requestVerifyEmailLink).not.toBeCalled();
      expect(setSecurity).toBeCalledWith('sessionTime', expectDate);
    });
  });
});
