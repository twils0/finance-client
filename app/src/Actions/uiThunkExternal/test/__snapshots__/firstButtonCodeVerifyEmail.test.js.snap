// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`uiThunkExternal firstButtonCodeVerifyEmail fails and throws an error when missing 'history' key in payload 1`] = `[Error: Please enter a value for the 'match' key - {}]`;

exports[`uiThunkExternal firstButtonCodeVerifyEmail fails and throws an error when missing 'match' key in payload 1`] = `[Error: Please enter a value for the 'match' key - {"history":{}}]`;
